File: Domain/Shared/ValueObject.cs
namespace Shared;

public abstract class ValueObject<T> where T : ValueObject<T>
{
    protected abstract IEnumerable<object> GetAttributesToIncludeInEqualityCheck();
    public override bool Equals(object other) => Equals(other as T);
    public bool Equals(T other)
    {
        if (other == null) 
            return false;

        return GetAttributesToIncludeInEqualityCheck().SequenceEqual(other.GetAttributesToIncludeInEqualityCheck());
    }
    public static bool operator ==(ValueObject<T> left, ValueObject<T>? right)
    {
        return Equals(left, right);
    }
    public static bool operator !=(ValueObject<T> left, ValueObject<T> right)=> !(left == right);

    public override int GetHashCode()
    {
        var hash = 13;
        var i = 1;
        foreach (var obj in this.GetAttributesToIncludeInEqualityCheck())
        {
            hash +=  (31 ^ i) * (obj == null ? 1 : obj.GetHashCode());
            i++;
        }
        return hash;
    }
}

File: java/shared/valueobject.java
package shared;

import java.util.Arrays;
import java.util.Objects;

public abstract class ValueObject<T extends ValueObject<T>> {
protected abstract Iterable<Object> getAttributesToIncludeInEqualityCheck();

@Override
public boolean equals(Object other) {
    if (this == other) return true;
    if (other == null || getClass() != other.getClass()) return false;
    T that = (T) other;
    return Arrays.equals(getAttributesToIncludeInEqualityCheck(), that.getAttributesToIncludeInEqualityCheck());
}

public static <T extends ValueObject<T>> boolean equals(ValueObject<T> left, ValueObject<T> right) {
    return Objects.equals(left, right);
}

@Override
public int hashCode() {
    int hash = 13;
    int i = 1;
    for (Object obj : getAttributesToIncludeInEqualityCheck()) {
        hash += (31 ^ i) * (obj == null ? 1 : obj.hashCode());
        i++;
    }
    return hash;
}
}